### PR TITLE
[Ready for Review] Upload files(s) on new Project

### PR DESCRIPTION
**Pull request by kushG**
<h3> Purpose </h3>
Currently, Onboarder allows a user to either create a new project or upload file(s) on an existing project. There's no way by which the user can create a new project & upload file(s) in that same project. This PR allows uploading file(s) on a newly created project.

Resolves https://github.com/CenterForOpenScience/osf.io/issues/2146

<h3> Changes </h3>
Adds couple of more VM javascript functions to handle project creation inside OBUploaderViewModel. project_new_post view function updated to return node serialized data.

<h3> Side Effects </h3>
View function modification might have missed an edge case

### CODE REVIEW

File: website/project/views/node.py [hunk 1]
```diff
@@ -78,9 +78,10 @@ def project_new_post(auth, **kwargs):
     user = auth.user
 
     title = strip_html(request.json.get('title'))
-    template = request.json.get('template')
-    description = strip_html(request.json.get('description'))
+    template = request.json.get('template', '')
```
Comment (sloria) [inline website/project/views/node.py]: Use `request.get_json()`; `request.json` is deprecated.

File: website/project/views/node.py [hunk 2]
```diff
@@ -102,9 +103,11 @@ def project_new_post(auth, **kwargs):
 
     else:
         project = new_node('project', title, user, description)
+        new_project = _view_project(project, auth)
 
     return {
-        'projectUrl': project.url
+        'projectUrl': project.url,
+        'new_node': new_project['node'] if new_project else None
```
Comment (sloria) [inline website/project/views/node.py]: For consistency with the other variables in the response, make `new_node` camelCase.

**Comment by sloria** [comment]
The Enter button does not submit the "Create" form. Can you look into this?

File: website/static/js/onboarder.js [hunk 3]
```diff
@@ -444,18 +447,18 @@ function OBUploaderViewModel(params) {
         self.target(selected);
         self.clearMessages();
         self.showProgress(true);
-        self.dropzone.options.url = function(files) {
+        self.dropzone.options.url = function (files) {
```
Comment (sloria) [inline website/static/js/onboarder.js]: Why all these added spaces?

**Comment by sloria** [comment]
I'm thinking that the "Upload to a new project" input and button should be hidden when an existing project is selected.

osf___dashboard

File: website/static/js/onboarder.js [hunk 4]
```diff
@@ -574,8 +576,43 @@ function OBUploaderViewModel(params) {
...
+
+    self.createSuccess = function (response) {
+        var node = serializeNode(response.new_node);
+        self.startUpload(node, self.selectedComponent, node.title, '');
+        if ((self.dropzone.getUploadingFiles().length) && (!self.dropzone.getQueuedFiles().length))
+            window.location = response.projectUrl;
+    };
+
+    self.createFailure = function () {
+        $osf.growl('Could not create a new project.', 'Please try again. If the problem persists, email <a href="mailto:support@osf.io.">support@osf.io</a>');
```
Comment (sloria) [inline website/static/js/onboarder.js]: > Use RavenJS (a JS client for Sentry) to log unexpected errors to our Sentry server.

https://cosdev.readthedocs.org/en/latest/osf/guidelines.html#client-side

File: website/templates/components/dashboard_templates.mako [hunk 5]
```diff
@@ -118,7 +119,28 @@
...
+                        data-bind="value: createNewUpload"
+                        placeholder="Enter a Project Name"
+                    >
+                </div>
+            </div>
+            <div data-bind="html: message(), attr: {class: messageClass()}" ></div>
+            <div class="row">
+                    <div class="col-md-12">
+                        <br>
+##                            <span data-bind="visible: showSubmit()">
```
Comment (sloria) [inline website/templates/components/dashboard_templates.mako]: Remove dead code and fix indentation here.

File: tests/test_views.py [hunk 6]
```diff
@@ -3840,6 +3840,16 @@ def test_creates_a_project(self):
...
 
+    def test_new_project_returns_serialized_node_data(self):
+        payload = {
+            'title': 'Im a real title'
+        }
+        res = self.app.post_json(self.url, payload, auth=self.creator.auth)
+        assert_equal(res.status_code, 201)
+        node = res.json['new_node']
+        assert_true(node)
+        assert_true(node['title'], 'Im a real title')
```
Comment (sloria) [inline tests/test_views.py]: This should be `assert_equal`

File: website/static/js/onboarder.js [hunk 7]
```diff
@@ -574,8 +576,43 @@ function OBUploaderViewModel(params) {
...
+//        return self.createNewUpload()!== null;
+//    });
+
+    self.submitCreateNewUpload = function () {
+        if (self.createNewUpload().trim() === '') {
+            self.changeMessage('Project Name is required', 'text-danger');
+            return false;
+        }
+        else {
+            $osf.postJSON(
```
Comment (sloria) [inline website/static/js/onboarder.js]: Methods that make requests should return promises, so return this XHR object.

File: website/static/js/onboarder.js [hunk 8]
```diff
@@ -415,6 +417,7 @@ function OBUploaderViewModel(params) {
     self.enableUpload = ko.observable(true);
     self.filename = ko.observable('');
     self.iconSrc = ko.observable('');
+    self.createNewUpload = ko.observable(null);
```
Comment (sloria) [inline website/static/js/onboarder.js]: This is a bit of a misnomer. Perhaps rename to `newProjectName`?

File: website/static/js/onboarder.js [hunk 9]
```diff
@@ -574,8 +576,43 @@ function OBUploaderViewModel(params) {
...
-}
 
+
+//    self.showSubmit = ko.computed(function(){
+//        return self.createNewUpload()!== null;
+//    });
+
+    self.submitCreateNewUpload = function () {
+        if (self.createNewUpload().trim() === '') {
+            self.changeMessage('Project Name is required', 'text-danger');
```
Comment (sloria) [inline website/static/js/onboarder.js]: Lowercase 'name'

File: website/templates/components/dashboard_templates.mako [hunk 10]
```diff
@@ -118,7 +119,28 @@
                 </osf-project-search>
             </div>
         </div><!-- end row -->
-        <div data-bind="html: message(), attr: {class: messageClass()}" ></div>
+        <form data-bind="submit: submitCreateNewUpload">
+            <div class="row">
+                <div class="col-md-12">
+                    <h4> "OR" Upload in a New Project</h4>
```
Comment (sloria) [inline website/templates/components/dashboard_templates.mako]: Change to

> OR Upload to a New Project

**Comment by sloria** [comment]
Pass finished. :point_up:

**Comment by samanehsan** [comment]
I'm going to bring this up to date and submit it as a new PR.

**Comment by samanehsan** [comment]
Continued in #2985

**Pull request closed without merging**